app/models/king.rb

```diff
@@ -1,7 +1,55 @@
 class King < Piece
+  def valid_move?(to_row, to_col)
+    logger.debug "*** KING: #{id} #{type} valid_move?"
+    return false if move_nil?(to_row, to_col)
+    return false if move_out_of_bounds?(to_row, to_col)
+    return false if move_destination_ally?(to_row, to_col)
+    return false unless move_legal?(to_row, to_col)
+    true
+  end
+
   def move_legal?(to_row, to_col)
+    logger.debug "move_legal for #{type}"
     row_diff = (row - to_row).abs
     col_diff = (col - to_col).abs
-    row_diff <= 1 && col_diff <= 1
+    return true if row_diff <= 1 && col_diff <= 1
+
+    legal_castle_move(to_row, to_col)
+  end # end move_legal?
+
+  private
+
+  def legal_castle_move(to_row, to_col)
+    return false unless legal_king_move?(to_row, to_col)
+    legal_rook_move?(to_col)
+  end
+
+  def legal_king_move?(to_row, to_col)
+    return false unless moves.empty?
+    return false unless row == to_row
+    return false unless (col - to_col).abs == 2
+    true
+  end
+
+  def legal_rook_move?(to_col)
+    rook = get_rook?(to_col)
+    return false unless rook
+    return false if move_obstructed?(row, rook.col)
```

Comment:
I thought that I checked that in rook_at at line 49. There is also a test where the rook has a move and prevents the castle.